framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,481 @@
+ * variable. Both type variables and wildcards may be missing a primary annotation.  For this set
+ * of annotations, there is an additional constraint that only wildcards that extend type
+ * variables can be missing annotations because
+ *
+ * In order to compute {@link #leastUpperBound(CFAbstractValue)} and
+ * {@link #mostSpecific(CFAbstractValue, CFAbstractValue)}, the case where one value has an
+ * annotation in a hierarchy and the other does not must be handled.  For type variables, the
+ * {@link AnnotatedTypeVariable} for the declaration of the type variable is used.  The
+ * {@link AnnotatedTypeVariable} is computed using the type mirror.  For wildcards, it is not
+ * always possible to get the {@link AnnotatedWildcardType} for the type mirror.  However, a
+ * CFAbstractValue's type mirror is only a wildcard if the type of some expression is a wildcard,
+ * so it is always the use of a wildcard rather than the declaration. For uses of wildcards, only
+ * the primary annotation on the upper bound is ever used.  So, the set of annotations represents
+ * the primary annotation on the wildcard's upper bound.  If that upper bound is a type variable,
+ * then the set of annotations could be missing an annotation in a hierarchy.
  *
+ * @author Stefan Heule
  */
 public abstract class CFAbstractValue<V extends CFAbstractValue<V>> implements AbstractValue<V> {
 
     /**
-     * The analysis class this store belongs to.
+     * The analysis class this value belongs to.
      */
     protected final CFAbstractAnalysis<V, ?, ?> analysis;
 
     protected final TypeHierarchy typeHierarchy;
+    protected final GenericAnnotatedTypeFactory<V, ?, ?, ?> factory;
+    protected final ProcessingEnvironment processingEnv;
+    protected final QualifierHierarchy hierarchy;
+    protected final TypeMirror underlyingType;
+    protected final Set<AnnotationMirror> annotations;
+
+    public CFAbstractValue(
+            CFAbstractAnalysis<V, ?, ?> analysis,
+            Set<AnnotationMirror> annotations,
+            TypeMirror underlyingType) {
+        this.analysis = analysis;
+        this.typeHierarchy = analysis.getTypeHierarchy();
+        this.annotations = annotations;
+        this.underlyingType = underlyingType;
+        this.factory = analysis.getTypeFactory();
+        this.processingEnv = factory.getProcessingEnv();
+        this.hierarchy = analysis.getTypeFactory().getQualifierHierarchy();
 
-    /**
-     * The type (with annotations) corresponding to this abstract value.
-     */
-    protected final AnnotatedTypeMirror type;
+        validateSet(this.getAnnotations(), this.getUnderlyingType());
+    }
 
     public CFAbstractValue(CFAbstractAnalysis<V, ?, ?> analysis, AnnotatedTypeMirror type) {
-        assert AnnotatedTypes.isValidType(analysis.qualifierHierarchy, type)
-                : "Encountered invalid type: " + type.toString(true);
+        if (type.getKind() == TypeKind.WILDCARD) {
```

Comment:
`@A T extends @B Object`
`? extends T'

Calling type.getEffectiveAnnotations() on the wildcard would return `@B`, this is not precise enough for this case:
````
<T> T getValue(List<? extends T> list) {
  T t = list.get(0)
  return t;
}
````

Instead the CFValue should store an empty set in case the wildcard extends a type variable.